registrations/permissions.py

```diff
@@ -0,0 +1,67 @@
+from django.contrib.auth.models import AnonymousUser
+from django.utils.translation import gettext as _
+from rest_framework import permissions
+from rest_framework.exceptions import PermissionDenied
+from rest_framework.request import Request
+
+from events.auth import ApiKeyUser
+
+
+class AuthenticatedGet(permissions.BasePermission):
+    def has_permission(self, request, view):
+        # Only authenticated users can get object
+        if request.method == "GET":
+            user = request.user
+
+            if isinstance(user, AnonymousUser):
+                return False
+
+        return super().has_permission(request, view)
+
+    def has_object_permission(self, request: Request, view, obj):
+        user = request.user
+
+        if request.method == "GET":
+            if not obj.can_be_edited_by(request.user):
+                return False
+
+            if isinstance(user, ApiKeyUser):
+                user_data_source, _ = view.user_data_source_and_organization
+                # allow updating only if the api key matches instance data source
+                if obj.data_source != user_data_source:
+                    return False
+
+        return True
```

Comment:
Could simplify it a bit, something like this (feel free to combine conditions or make return values more explicit):
```suggestion
    def has_object_permission(self, request: Request, view, obj):
        if request.method != "GET":
            return True

        if obj.can_be_edited_by(request.user):
            return True

        if isinstance(request.user, ApiKeyUser):
            user_data_source, _ = view.user_data_source_and_organization
            # allow updating only if the api key matches instance data source
            return obj.data_source == user_data_source:

        return True
```

Generally speaking it's a good idea to avoid negations to make the code more readable. For example, "editable by x" is easier to process than "not editable by x". This is not always the case, of course, depends on the context. E.g. the short circuit conditions in my suggestion return True, so one returning False and the other returning True could be confusing.

Reply:
I edited the suggestion little but and end result is:
```
def has_object_permission(self, request: Request, view, obj):
        user = request.user

        if request.method != "GET":
            return True

        if isinstance(user, ApiKeyUser):
            user_data_source, _ = view.user_data_source_and_organization
            # allow to view data only if the api key matches instance data source
            if obj.data_source != user_data_source:
                return False
        
        return obj.can_be_edited_by(request.user)
```

The purpose for the change is to cover a case that api key user has permissions to the organization but the data source is incorrect. Added also a unit test to cover this case.